File: ac_topic_subscription_cpp/src/oopsubscription.cpp
#include<rclcpp/rclcpp.hpp>
#include<std_msgs/msg/int32.hpp>
#include<geometry_msgs/msg/twist.hpp>
#include<nav_msgs/msg/odometry.hpp>

class Subscription: public rclcpp::Node
{
    public:
        // Initiate a Node called Subscription
        Subscription():Node("Subscription")
        {
            // Create a Subscriber object that will listen to the /counter topic and will call the 'topic_callback' function each time it reads something
            //10 is the subscription queue size. you can also create a qos_profile variable and create Qos object and pass it in instead of just one number!
            sub_ = this->create_subscription<std_msgs::msg::Int32>("/counter",10, std::bind(&Subscription::IntCallback , this , std::placeholders::_1)); 
            sub2_ = this->create_subscription<geometry_msgs::msg::Twist>("/cmd_vel" , 10 , std::bind(&Subscription::TwistCallback , this , std::placeholders::_1));
            sub3_ = this->create_subscription<nav_msgs::msg::Odometry>("/odom" , 10 , std::bind(&Subscription::OdomCallback ,this , std::placeholders::_1 ));
        }


    private:
        // Define a function called 'topic_callback' that receives a parameter named 'msg'
        void IntCallback(const std_msgs::msg::Int32::SharedPtr msg)
        {
            // Print the value 'data' inside the 'msg' parameter
            RCLCPP_INFO(this->get_logger() , "Hey: '%d'" , msg->data);
        }

        void TwistCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
        {
            RCLCPP_INFO(this->get_logger() , "linear_x: %f  , angular_z: %f ", msg->linear.x ,msg->angular.z );
        }

        void OdomCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
        {
            RCLCPP_INFO(this->get_logger() , "x: %f ,y: %f ,z: %f", msg->pose.pose.position.x , msg->pose.pose.position.y , msg->pose.pose.position.z);
        }

        rclcpp::Subscription<std_msgs::msg::Int32>::SharedPtr sub_;
        rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub2_;
        rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub3_;
};

int main(int argc , char* argv[])
{
    rclcpp::init(argc , argv); 
    auto node = std::make_shared<Subscription>();
    // Create a loop that will keep the program in execution
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0 ;
}

File: ac_topic_subscription_py/ac_topic_subscription_py/oopsubscription.py
import rclpy
# import the ROS2 python libraries
from rclpy.node import Node
from std_msgs.msg import Int32
from geometry_msgs.msg import Twist
from nav_msgs.msg import Odometry
# import the LaserScan module from sensor_msgs interface
from sensor_msgs.msg import LaserScan
# import Quality of Service library, to set the correct profile and reliability in order to read sensor data.
from rclpy.qos import ReliabilityPolicy , QoSProfile #https://docs.ros2.org/latest/api/rclpy/api/qos.html

class Subscription(Node):
    def __init__(self):
        super().__init__('Subscription')
        self.sub_ = self.create_subscription(Int32 , "counter" , self.IntCallback , QoSProfile(depth=10 , reliability = ReliabilityPolicy.BEST_EFFORT))
        self.sub2_ = self.create_subscription(Twist ,"cmd_vel" ,self.TwistCallback , QoSProfile(depth=10 , reliability = ReliabilityPolicy.BEST_EFFORT))
        self.sub3_ = self.create_subscription(Odometry , "odom", self.OdomCallback , QoSProfile(depth=10 , reliability = ReliabilityPolicy.BEST_EFFORT))
        self.sub4_ = self.create_subscription(LaserScan ,"scan" ,  self.LaserCallback, QoSProfile(depth=10 , reliability = ReliabilityPolicy.BEST_EFFORT)) #is the most used to read LaserScan data and some sensor data too.
        
        self.sub_ # prevent unused variable warning
        self.sub2_ # prevent unused variable warning
        self.sub3_ # prevent unused variable warning
        self.sub4_ # prevent unused variable warning

        # define the variable to save the received info
        self.laser_data = 0

    def IntCallback(self , msg):
        self.get_logger().info("IntCallback: '%s'" % str(msg))
        
    def TwistCallback(self , msg):        
        self.get_logger().info("TwistCallback: linear=[%f , %f ,  %f]   "  %(msg.linear.x , msg.linear.y , msg.linear.z) )

    def OdomCallback(self , msg):
        self.get_logger().info("OdomCallback: '%s'  "  %str(msg))

    def LaserCallback(self , msg):
        self.get_logger().info("LaserCallback: '%s'  "  %str(msg))
        # save the received data
        self.laser_data = msg.ranges[0]
        # print the received data
        print(self.laser_data)


def main(args=None):
    rclpy.init(args=args)
    node = Subscription()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__init__':
    main()